### PR TITLE
implementation of waitFor for the node definition

### PR DESCRIPTION
**Pull request by steiler**
This PR holds the implementation of the waitFor config stanza.

Users can define specific create / start dependencies between the config defined nodes.

**Sample topology**
```
name: waitForTest
topology:
  kinds:
    srl:
      type: ixrd3
      image: ghcr.io/nokia/srlinux
    linux:
      image: ghcr.io/hellt/network-multitool
  nodes:
    srl1:
      kind: srl
    srl2:
      kind: srl
    srl3:
      kind: srl
      waitFor:
        - srl1
        - srl2
    linux:
      kind: linux
      waitFor:
        - srl3
```
**Relevant Output:**
```
DEBU[0001] Dependencies:
srl2 -> [  ]
srl3 -> [ srl2, srl1 ]
linux -> [ srl3 ]
srl1 -> [  ] 
DEBU[0001] - cycle check round 1 - 
linux <- [  ]
srl1 <- [ srl3 ]
srl2 <- [ srl3 ]
srl3 <- [ linux ] 
DEBU[0001] - cycle check round 2 - 
srl2 <- [ srl3 ]
srl3 <- [  ]
srl1 <- [ srl3 ] 
DEBU[0001] - cycle check round 3 - 
srl1 <- [  ]
srl2 <- [  ] 
DEBU[0001] node creation graph is successfully validated as being acyclic 
```

### CODE REVIEW

**Comment by steiler** [comment]
Adding waitFor also for the Kind level.
For the default it will not make sense.

**Config:**
```
# topology documentation: http://containerlab.dev/lab-examples/single-srl/
name: waitForTest
topology:
  kinds:
    srl:
      type: ixrd3
      image: ghcr.io/nokia/srlinux
    linux:
      image: ghcr.io/hellt/network-multitool
      waitFor:
        - srl1
        - srl2
        - srl3
  nodes:
    srl1:
      kind: srl
    srl2:
      kind: srl
    srl3:
      kind: srl
      waitFor:
        - srl1
        - srl2
    linux:
      kind: linux
```

**Relevant Output:**
```
DEBU[0004] Dependencies:
srl2 -> [  ]
srl3 -> [ srl1, srl2 ]
linux -> [ srl1, srl2, srl3 ]
srl1 -> [  ] 
DEBU[0004] - cycle check round 1 - 
srl1 <- [ linux, srl3 ]
srl2 <- [ linux, srl3 ]
srl3 <- [ linux ]
linux <- [  ] 
DEBU[0004] - cycle check round 2 - 
srl1 <- [ srl3 ]
srl2 <- [ srl3 ]
srl3 <- [  ] 
DEBU[0004] - cycle check round 3 - 
srl2 <- [  ]
srl1 <- [  ] 
DEBU[0004] node creation graph is successfully validated as being acyclic 
```

File: types/types.go [hunk 1]
```diff
@@ -139,7 +139,8 @@ type NodeConfig struct {
 	// status that is set by containerlab to indicate deployment stage
 
 	// Extras
-	Extras *Extras `json:"extras,omitempty"` // Extra node parameters
+	Extras  *Extras  `json:"extras,omitempty"` // Extra node parameters
+	WaitFor []string `json:"waitFor,omitempty"`
```
Comment (hellt) [inline types/types.go:143]: json and yaml tags for composite words should use dashes
`waitFor` -> `wait-for`

This is aligned with the rest of the fields we expose in clab config

File: types/types.go [hunk 2]
```diff
@@ -182,11 +183,13 @@ func (node *NodeConfig) GenerateConfig(dst, templ string) error {
...
 	}
-	defer f.Close()
 
 	_, err = f.Write(dstBytes.Bytes())
+	if err != nil {
+		return err
+	}
 
-	return err
+	return f.Close()
```
Comment (hellt) [inline types/types.go:192]: what is the reason of removing `defer`? If `f.Write` errors, but the file has already been opened, then it won't be closed without a defer.

File: types/topology.go [hunk 3]
```diff
@@ -405,7 +405,7 @@ func (t *Topology) GetNodeMemory(name string) string {
 	return ""
 }
 
-// Return the Sysctl configuration for the given node.
+// GetSysCtl Return the Sysctl configuration for the given node.
```
Comment (hellt) [inline types/topology.go:408]: thanks for fixing the comments patter
could you also decapitalise the next word after the method name? `Return->return`

File: types/topology.go [hunk 4]
```diff
@@ -416,7 +416,7 @@ func (t *Topology) GetSysCtl(name string) map[string]string {
 	return nil
 }
 
-// Return the Subject Alternative Name configuration for the given node.
+// GetSANs Return the Subject Alternative Name configuration for the given node.
```
Comment (hellt) [inline types/topology.go:419]: thanks for fixing the comments patter
could you also decapitalise the next word after the method name?

File: types/topology.go [hunk 5]
```diff
@@ -426,7 +426,7 @@ func (t *Topology) GetSANs(name string) []string {
 	return nil
 }
 
-// Returns the 'extras' section for the given node.
+// GetNodeExtras Returns the 'extras' section for the given node.
```
Comment (hellt) [inline types/topology.go:429]: thanks for fixing the comments patter
could you also decapitalise the next word after the method name?

File: types/node_definition.go [hunk 6]
```diff
@@ -289,6 +291,13 @@ func (n *NodeDefinition) GetSANs() []string {
 	return n.SANs
 }
 
+func (n *NodeDefinition) GetWaitFor() []string {
```
Comment (hellt) [inline types/node_definition.go:294]: add a comment?

File: docs/manual/nodes.md [hunk 7]
```diff
@@ -535,6 +535,63 @@ topology:
         net.ipv6.icmp.ratelimit: 1000
 ```
 
+### waitFor
```
Comment (hellt) [inline docs/manual/nodes.md:538]: change to `wait-for`

File: docs/manual/nodes.md [hunk 8]
```diff
@@ -535,6 +535,63 @@ topology:
...
+For the explicit definition of startup dependencies between nodes, the `waitFor` knob under the `kind` or `node` level can be used.
+
+Node _srl3_ will wait until _srl1_ and _srl2_ are created before it is created. The _linuxNode_ will via the definition in the _linux_ kind wait for all three _srl_ nodes to be created before it gets created.
```
Comment (hellt) [inline docs/manual/nodes.md:541]: can you change `linuxNode` to `client`. Makes it easier to read I think

File: schemas/clab.schema.json [hunk 9]
```diff
@@ -258,6 +258,14 @@
...
                 "config": {
                     "$ref": "#/definitions/config-config"
+                },
+                "waitFor": {
+                    "type": "array",
+                    "items": {
+                        "type": "string"
+                    },
+                    "uniqueItems": true,
+                    "description": "Define which nodes should be started befor this node will start"
```
Comment (hellt) [inline schemas/clab.schema.json:268]: can you also add

```
"markdownDescription": "[wait-for](https://containerlab.dev/manual/nodes/#cmd) defines which nodes should be started before this node will start"
```

Comment (hellt) [reply schemas/clab.schema.json:268]: also a typo in `before` word

File: types/types.go (hunk 2 above)
Comment (steiler) [reply types/types.go:192]: This was not my free will but Deepsource complaining.
Here is the reference: https://deepsource.io/gh/srl-labs/containerlab/run/7183e662-b504-4b99-a10c-3047bf5473f6/go/GO-S2307?listindex=0

Deepsource seems to only inspect files or sections of files touched in the related commits. Thats why this "issue" popped up although I did not even mean to touch this part in first place.
So my changes in this case only follow the Deepsource recommendations.

File: types/node_definition.go (hunk 6 above)
Comment (steiler) [reply types/node_definition.go:294]: None of the Getters in that file has any comment. And it is basically just a getter...
https://github.com/srl-labs/containerlab/blob/1eb93a1382932905272e3a237120cf8d35dd3b91/types/node_definition.go#L294
If you want a comment, we should add that to all of the getters there

File: types/topology.go (hunk 3 above)
Comment (steiler) [reply types/topology.go]: done!

File: types/types.go (hunk 1 above)
Comment (steiler) [reply types/types.go]: done!

File: types/topology.go (hunk 4 above)
Comment (steiler) [reply types/topology.go]: done!

File: types/topology.go (hunk 5 above)
Comment (steiler) [reply types/topology.go]: done!

File: docs/manual/nodes.md (hunk 7 above)
Comment (steiler) [reply docs/manual/nodes.md]: done!

File: docs/manual/nodes.md (hunk 8 above)
Comment (steiler) [reply docs/manual/nodes.md]: done!

File: schemas/clab.schema.json (hunk 9 above)
Comment (steiler) [reply schemas/clab.schema.json]: done!

**Pull request merged**